src/CodeGen_Xtensa.cpp

```diff
@@ -195,7 +202,13 @@ void CodeGen_Xtensa::compile(const LoweredFunc &f, const std::map<std::string, s
             UsesDmaCopy uses_dma;
             body.accept(&uses_dma);
             if (uses_dma.uses_dma) {
-                stream << "ScopedDmaInitializer dma_initializer;\n";
+                stream << get_indent() << "ScopedDmaInitializer dma_initializer(" << uses_dma.max_channel_no + 1 << ");\n";
+                stream << get_indent() << "if (!dma_initializer.is_valid()) {\n";
+                stream << get_indent() << "halide_error("
```

Comment:
Random thought: maybe generate a local function of the form

int halide_error_dma_failure(void* user_context) {
    halide_error(user_context, "DMA failure");
    return halide_error_code_generic_error;
}

then you can make the failure clause simpler, and (possibly) slightly less code.